Fix the sorting mechanism on the Stock page

marsaldev opened the pull request:
<!-----------------------------------------------------------------------------
Thank you for contributing to the PrestaShop project! 

Please take the time to edit the "Answers" rows below with the necessary information.

Check out our contribution guidelines to find out how to complete it:
https://devdocs.prestashop.com/8/contribute/contribution-guidelines/#pull-requests
------------------------------------------------------------------------------>

| Questions         | Answers
| ----------------- | -------------------------------------------------------
| Branch?           | develop
| Description?      | Fix wording in Stock page.
| Type?             | bug fix
| Category?         |  BO
| BC breaks?        | no
| Deprecations?     | no
| Fixed ticket?     | Fixes #9629.
| Related PRs       | ~
| How to test?      | ~
| Possible impacts? | nope


<!-- Click the form's "Preview" button to make sure the table is functional in GitHub. Thank you! -->

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/prestashop/prestashop/28140)
<!-- Reviewable:end -->

Commit:
Update wording for filtering

Julievrz commented:
Hi,

I'm sorry but I don't understand this new wording. Actually, I don't understand the issue at all:

In 1.7.8.5, when I click on the arrow next to ID, the sorting looks correct. Same thing if I click on the arrow next to the Product name, the sorting looks like A > B. 

I think I'm missing something here. 
![image](https://user-images.githubusercontent.com/70583503/161920269-0473df4b-d04e-4e22-be91-21943cfd6714.png)
![image](https://user-images.githubusercontent.com/70583503/161920323-ad73b513-b8ee-44db-8be1-621d543e53ec.png)

marsaldev commented:
Hi @Julievrz 😊
I admit that to understand the issue it took me a bit 😊
In the Stock page (Catalog > Stock) the ordering by Product is not alphabetical, but by ID product (and combination), so is different from the products list page.
Your ss is from products list, where the products can be order alphabetically

Julievrz commented:
> Hi @Julievrz 😊 I admit that to understand the issue it took me a bit 😊 In the Stock page (Catalog > Stock) the ordering by Product is not alphabetical, but by ID product (and combination), so is different from the products list page.
> 
> Your ss is from products list, where the products can be ordered alphabetically

Oh ok, I see. Thank you. Wouldn't it be clearer to add the ID column and put the arrow on this one instead of on the Product column?

marsaldev commented:
> > Hi @Julievrz 😊 I admit that to understand the issue it took me a bit 😊 In the Stock page (Catalog > Stock) the ordering by Product is not alphabetical, but by ID product (and combination), so is different from the products list page.
> > Your ss is from products list, where the products can be ordered alphabetically
> 
> Oh ok, I see. Thank you. Wouldn't it be clearer to add the ID column and put the arrow on this one instead of on the Product column?

In that case we need to change also the sorting mechanism of "Product" column 🤔
Your proposal is to add the ID column that acquire the actual sorting of "Product" and fix the sorting of "Product" that will sort by name (like the Products list page)?

kpodemski commented:
Hello @marsaldev 

This issue is about:

> If I'm not mistaken, there is no wording issue here but a bug: on the Catalog > Stocks pages, when the user is using the product filter, he expects the listing to be reorganized by alphabetical order (from A to Z), and this is how it is supposed to work. Here, he gets a reorganization by numerical order (depending on the product's ID). In short, the filter parameter needs to be changed to fix this bug. @khouloudbelguith, do you confirm?

So the goal would be to change the sorting mechanism to:
- sort by an actual product name
- and as suggested above, add a possibility to sort by `id` too

👍

kpodemski reviewed (changes requested):
I'm blocking this PR to ensure it's not merged before changes.

Commit:
Add filtering by ID product

marsaldev commented:
Applied edits to allow user sort by Product ID, and changed the default sort by Product ID (descend, from higher value to lowest).
Updated the wording for the "Product" name, now is "Product (name)", I guess is more clear to who want to sort by name.
It miss one last thing, after struggling with Vue (where my knowledge is very basic), the arrow sort icon at first visit is "wrong", I don't know how to set it. (See ss)
<img width="125" alt="Schermata 2022-04-08 alle 15 55 45" src="https://user-images.githubusercontent.com/48283112/162450396-04931900-f879-4d12-b21f-1be97335ab7e.png">
 
In first instance I've also added the column "Combination ID", so in the PHP files you'll find also references to the combination, but I've removed it (in the Vue files) to no add another column and create confusion, but let me know if we need to add it also.

Julievrz commented on admin-dev/themes/new-theme/js/app/pages/stock/components/overview/products-table.vue:
```diff
@@ -26,13 +26,24 @@
+            @sort="sort"
+            :current-sort="currentSort"
+          >
+            {{ trans('title_product_id') }}
```
Is "title_product_id" a translatable string?

marsaldev replied:
Yep 😊

Julievrz replied:
What would be the translation domain? If you show me a screenshot I can tell you the right domain. :-)

marsaldev replied:
You've suggested a change, just commented 😊

Julievrz replied:
I know :D But the string doesn't seem to be in any domain.

marsaldev replied:
@Julievrz 
Mmm I think I don't get you 😊 
The string is this this one: https://github.com/PrestaShop/PrestaShop/pull/28140#discussion_r847275044
And the domain proposed by you is _Admin.Global_, but I'm not sure about that, maybe we can keep _Admin.Catalog.Feature_ or I'd propose _Admin.Catalog.Stock_

Here you are the screen of the table: 
<img width="1678" alt="Schermata 2022-04-12 alle 11 36 10" src="https://user-images.githubusercontent.com/48283112/162930126-4dc373ef-f129-4f17-9df3-308d1f6806e2.png">

Julievrz replied:
Sorry, I'll try to rephrase 😇 

The Admin.Catalog.Stock doesn't exist. We could use the Admin.Catalog.Feature (see https://devdocs.prestashop.com/1.7/development/internationalization/translation/translation-domains/). I'm not sure we'll be able to be super specific as this string will probably appear on the product page, which is in Admin.Catalog as well... 😅

But I was talking about this string: 
![image](https://user-images.githubusercontent.com/70583503/162931772-a5d3ae4b-f090-4c44-9590-f7ad753b73b8.png) which doesn't seem to have any domain.

marsaldev replied:
@Julievrz
> Sorry, I'll try to rephrase 😇

No worries! Thanks for the document! Didn't aware of it 😊 
And we cannot create a new domain if needed?

> But I was talking about this string:
> ![image](https://user-images.githubusercontent.com/70583503/162931772-a5d3ae4b-f090-4c44-9590-f7ad753b73b8.png) which doesn't seem to have any domain.

That string is a new string, that's why are you saying doesn't seem to have any domain?
The translation and the domain is in the `StockApi.php` file 😊 
I used the _Admin.Catalog.Feature_ but you suggest _Admin.Global_, so you're saying to apply _Admin.Global_ because the string doesn't have any domain yet?

Julievrz replied:
Ok, I understand now, I thought we were talking about 2 different strings. 🤯 😄 

The string 'ID' already exists in Admin.Global, that's why I told you to use Admin.Global, to avoid creating a similar string in another domain (because when you do that, that's more work for translators). The string 'ID' can appear on many pages of the back office, that's why using Admin.Global seemed like a good idea: translators will only have to translate it one time for the whole BO. 

However, if you think the translation might be different on the Stock page, you could be more specific and create a new string in Admin.Catalog.Feature.

I hope that helps.

marsaldev replied:
Oh ok, got it!
I think it should be better distinguish the translations for each page, but maybe it is very specific case 🤔
The fact is we have a lot of "Entity" and the `ID` is the base field for them, I don't remember in how much _Grid_ we show it, and for someone translating that one will translate also this one (in the Stock page), wdyt?

Julievrz commented on src/PrestaShopBundle/Translation/Api/StockApi.php:
```diff
@@ -70,14 +70,16 @@ public function getTranslations()
             'product_low_stock_alert' => $this->translator->trans('Low stock alert:', [], 'Admin.Catalog.Feature'),
             'title_available' => $this->translator->trans('Available', [], 'Admin.Global'),
             'title_bulk' => $this->translator->trans('Bulk edit quantity', [], 'Admin.Catalog.Feature'),
+            'title_combination_id' => $this->translator->trans('ID Combination', [], 'Admin.Catalog.Feature'),
```
```suggestion
            'title_combination_id' => $this->translator->trans('Combination's ID', [], 'Admin.Catalog.Feature'),
```

Julievrz commented on src/PrestaShopBundle/Translation/Api/StockApi.php:
```diff
@@ -70,14 +70,16 @@ public function getTranslations()
             'title_movements_type' => $this->translator->trans('Type', [], 'Admin.Global'),
             'title_physical' => $this->translator->trans('Physical', [], 'Admin.Catalog.Feature'),
-            'title_product' => $this->translator->trans('Product', [], 'Admin.Global'),
+            'title_product' => $this->translator->trans('Product (name)', [], 'Admin.Global'),
```
```suggestion
            'title_product' => $this->translator->trans('Product name', [], 'Admin.Global'),
```

Julievrz commented on src/PrestaShopBundle/Translation/Api/StockApi.php:
```diff
@@ -70,14 +70,16 @@ public function getTranslations()
             'title_physical' => $this->translator->trans('Physical', [], 'Admin.Catalog.Feature'),
-            'title_product' => $this->translator->trans('Product', [], 'Admin.Global'),
+            'title_product' => $this->translator->trans('Product (name)', [], 'Admin.Global'),
+            'title_product_id' => $this->translator->trans('ID', [], 'Admin.Catalog.Feature'),
```
```suggestion
            'title_product_id' => $this->translator->trans('ID', [], 'Admin.Global'),
```

marsaldev replied:
🤔  Maybe here it's better specify another domain? Or there'll be the risk to change the translations also in others page in the BO, wdyt?

Julievrz replied:
Yes, we could be more specific here. However, I'm not sure we'll be able to be super specific as this string will probably appear on the product page, which is in Admin.Catalog as well... 😅

Julievrz reviewed (commented):
A few comments about wording. Would you mind sharing a screenshot of the table, please? 😇

marsaldev committed:
Apply suggestions from code review

Co-authored-by: Julie Varisellaz <70583503+Julievrz@users.noreply.github.com>

Progi1984 commented on src/PrestaShopBundle/Translation/Api/StockApi.php:
```diff
@@ -70,14 +70,16 @@ public function getTranslations()
             'product_low_stock_alert' => $this->translator->trans('Low stock alert:', [], 'Admin.Catalog.Feature'),
             'title_available' => $this->translator->trans('Available', [], 'Admin.Global'),
             'title_bulk' => $this->translator->trans('Bulk edit quantity', [], 'Admin.Catalog.Feature'),
+            'title_combination_id' => $this->translator->trans('Combination's ID', [], 'Admin.Catalog.Feature'),
```
```suggestion
            'title_combination_id' => $this->translator->trans('Combination\'s ID', [], 'Admin.Catalog.Feature'),
```

Progi1984 reviewed (commented):
A small feedback

marsaldev committed:
Update src/PrestaShopBundle/Translation/Api/StockApi.php

Co-authored-by: Progi1984 <progi1984@gmail.com>

marsaldev committed:
Update src/PrestaShopBundle/Translation/Api/StockApi.php

Co-authored-by: Julie Varisellaz <70583503+Julievrz@users.noreply.github.com>

matks commented:
I re run CI

sLorenzini reviewed (commented):
hello @marsaldev,

I tested the PR and I have found issues:

On catalog > stock page
The arrow next to ID is on top so the sort should be asc. Products are sort by desc way, like the products page. 
We need to click twice on the arrow to have products sort by asc way.

Stock page:

https://user-images.githubusercontent.com/2854834/165519482-d34e12c5-86bc-431d-84ce-4dfb2492875e.mp4

Products page:

https://user-images.githubusercontent.com/2854834/165519766-016992f2-659f-4bce-8fab-7fb4b0aecfe3.mp4

Good on products page, develop and 178x branches.

----------------------------

On catalog > stock page
Click on arrow next to id
Products are sort by asc
Click on pagination page
The sort is not saved, keep the default sort of the page


https://user-images.githubusercontent.com/2854834/165520537-128728aa-ae84-4338-b84a-da71ffb366af.mp4

In products page all good, develop and 178x branches.

Could you please check?

Thanks

marsaldev commented:
Hi @sLorenzini, I know about that :(
I need help from a front end dev, because I can't figure out how to solve, maybe @atomiix ? 😊

marsaldev commented:
> I need help from a front end dev, because I can't figure out how to solve, maybe @atomiix ? 😊

Or @NeOMakinG 😊

Commit:
Change default directon of the product id filter

Commit:
Add sort to the store and use it on pagination

sLorenzini reviewed (approved):
Hello @marsaldev,

Tested on develop branch
7.2.21 PHP version 

Different cases of test:
- Tested on desktop screen the asc/desc way of sort + pagination, all good
- Tested on mobile screen the asc/desc way of sort + pagination, all good
- Update quantities and sort products, all good
- Add new products with images and no images, sort products in stock page by id, all good
- check some products and change quantities for all, all good

QA approved ✅ 

Thanks

Progi1984 commented:
Thanks @marsaldev & @sLorenzini
